tests/coverage/mcdc/match_misc.rs

```diff
@@ -0,0 +1,190 @@
+            )*
+            _ => say("not matched"),
+        }
+    }
+}
+    variant_identifier!(val, (0, 1, 2));
+}
+
+// No match pairs when lowering matching tree.
+fn empty_matching_decision(val: i32) {
+    match val {
+        x if x > 8 && x < 10 => say("in (8, 10)"),
+        x if x > 4 && x < 7 => say("in (4, 7)"),
+        _ => say("other"),
+    }
+}
+
+// Matching decision skips the first candidate
+fn skipped_matching_decision(val: i32) {
+    match val {
+        x if x >= 0 => say("non-negative"),
+        -1 => say("-1"),
+        _ => say("other"),
+    }
+}
+
+// The first two candidates share same condition.
+fn overlapping_decisions(val: (Option<i32>, Option<i32>)) {
+    match val {
+        (Some(_), Some(_)) => say("both some"),
+        (Some(_), None) | (None, Some(_)) => say("one and only one some"),
+        (None, None) => say("none"),
+    }
+}
+
+fn partial_matched_decision(val: u8) {
+    // `b'-'` is the second test while `b'0'..=b'9'` is the last, though they
+    // are in same candidate.
+    match val {
+        b'"' | b'r' => say("quote or r"),
+        b'0'..=b'9' | b'-' => say("number or -"),
+        b't' | b'f' => say("t or f"),
+        _ => {}
+    }
+}
+
+// Patterns are tested with several basic blocks.
+fn partial_matched_with_several_blocks(val: u8) {
+    match val {
+        b'a'..=b'f' => say("hex"),
+        b'A'..=b'F' => say("hex upper"),
+        b'_' => say("underscore"),
+        _ => say("break"),
+    }
+}
+
+fn match_failure_test_kind(val: bool, opt: Option<i32>) {
+    match (val, opt) {
+        (false, None) => say("none"),
+        (false, Some(_)) => say("some"),
+        _ => say("other"),
+    }
+}
+
+enum Pat {
+    A(i32),
+    B(i32),
+}
+
+// The last arm is shown like a condition but it never fails if tested.
+fn uncoverable_condition(val: (Pat, Pat)) {
+    match val {
+        (Pat::A(a), Pat::A(b)) => say(&(a + b).to_string()),
+        (Pat::B(a), _) | (_, Pat::B(a)) => say(&a.to_string()),
+    }
+}
+
+fn nested_matching(a: bool, val: Pat) {
+    if a && match val {
+        Pat::A(x) => x == 2,
+        _ => false,
+    } {
+        say("yes");
+    }
+}
+
+// It's possible to match two arms once.
+fn multi_matched_candidates(val: Pat, a: i32) {
+    match val {
+        Pat::A(f) if f == a => say("first"),
+        Pat::A(1) if a > 0 => say("second"),
+        _ => say("other"),
+    }
+}
+
+fn empty_subcandidate(val: Pat) {
+    match val {
+        Pat::A(1) | Pat::A(2) => say("first"),
+        // The first two condition in this pattern is redundant indeed.
+        // But this piece of code is legitimate and it could cause a subcandidate
+        // with no match pair.
+        Pat::A(_) | Pat::B(_) | _ => say("other"),
+    }
+}
+
+fn implicit_folded_condition(val: (bool, bool)) {
+    match val {
+        // The first `true` is always matched if tested.
+        (false, false) | (true, true) => say("same"),
+        _ => say("not same"),
+    }
+}
```

Comment:
The first `true` in such pattern should be considered as "constant folded". But llvm does not take condition whose only one branch is `CovTerm::Zero` as folded now. This can be solved if [#94317](https://github.com/llvm/llvm-project/pull/94137) was accepted.

Reply:
The second issue is llvm does not count `FalseCount` when calculating `MaxCounterID`, which is also fixed by [#94317](https://github.com/llvm/llvm-project/pull/94137). As far as I tested this can be more common in mcdc for pattern matching.